streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/TaskAssignor.java

```diff
@@ -52,12 +56,30 @@
     private final Set<TaskPair<T>> taskPairs;
     private final int maxNumTaskPairs;
     private final ArrayList<T> tasks;
+    private boolean sameTopology = true;
```

Comment:
A general comment is to have to variables here: `prevAssignmentBalanced` and `prevClientsUnchanged`, and then update the two variables such like:

```
prevAssignmentBalanced = prevAssignmentBalanced && oldTasks.size() < 2 * avgNumTasks && oldTasks.size() > avgNumTasks / 2;
```

And then use the same state only if

```
prevAssignmentBalanced && prevClientsUnchanged && state.prevAssignedTasks.contains(task)
```